### PR TITLE
Issue GRADLE-3349 (GroovyDoc task does not mark all properties that affect the output of the task as inputs)

### PR DESCRIPTION
**Pull request by Myllyenko**
The overview property of the Groovydoc class has been annotated as InputFile and Optional

### CODE REVIEW

**Comment by eljobe** [comment]
@Myllyenko Thanks for the contribution.

Please add an integration test to [GroovyDocIntegrationTest](https://github.com/gradle/gradle/blob/master/subprojects/language-groovy/src/integTest/groovy/org/gradle/groovy/GroovyDocIntegrationTest.groovy) which shows that when using the overview setting:

1. building initially
1. modifying the overview file
1. building again

Causes the outputs to change.  In other words, add a test that would have failed without your change, but which now passes.  Let me know if you need me to find you an example test for a different task for you to steal ideas from.

**Comment by big-guy** [comment]
@Myllyenko I went ahead and added a example test that I think will cover this case: 

https://github.com/gradle/gradle/blob/master/subprojects/language-groovy/src/integTest/groovy/org/gradle/groovy/GroovyDocIntegrationTest.groovy#L63

If you remove the `@NotYetImplemented`, you should be able to run `/gradlew languageGroovy:integTest --tests "*GroovyDocIntegrationTest"` to test it.

Thanks for the contribution!

**Comment by eljobe** [comment]
@big-guy Thanks for writing this test for @Myllyenko. You are the best. @Myllyenko, if you'll update this pull request by adding a commit which removes `@NotYetImplemented` from the test as @big-guy suggested, I will be able to continue with the review and merge of this change.

**Comment by Myllyenko** [comment]
Thank you, @big-guy!

Your test has exposed a bug in my contribution. Namely, it fails to cast that ```String``` property to ```File```: [https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/groovy/org/gradle/api/internal/project/taskfactory/InputFilePropertyAnnotationHandler.java#L29](url)

I've tried to circumvent this by adding the following method:
```
    @InputFile
    @Optional
    public File getOverviewFile() {
        return getProject().file(this.getOverview());
    }
```

But with this it fails with the exception when overview isn't specified:
```
Caused by: java.lang.IllegalArgumentException: Neither path nor baseDir may be null or empty string. path='null' basedir='C:\workspaces\git-hub\gradle\subprojects\language-groovy\build\tmp\test files\GroovyDocIntegrationTest\can_run_groovydoc\qlqqa'
	at org.gradle.api.internal.file.BaseDirFileResolver.doResolve(BaseDirFileResolver.java:67)
	at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:84)
	at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:66)
	at org.gradle.api.internal.file.DefaultFileOperations.file(DefaultFileOperations.java:77)
	at org.gradle.api.internal.project.AbstractProject.file(AbstractProject.java:660)
	at org.gradle.api.tasks.javadoc.Groovydoc.getOverviewFile(Groovydoc.java:284)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$PropertyInfo$4.create(AnnotationProcessingTaskFactory.java:457)
	at org.gradle.util.SingleMessageLogger.whileDisabled(SingleMessageLogger.java:166)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$PropertyInfo.getValue(AnnotationProcessingTaskFactory.java:455)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$Validator$1.call(AnnotationProcessingTaskFactory.java:264)
	at org.gradle.api.internal.file.collections.BuildDependenciesOnlyFileCollectionResolveContext.add(BuildDependenciesOnlyFileCollectionResolveContext.java:73)
	at org.gradle.api.internal.file.collections.BuildDependenciesOnlyFileCollectionResolveContext.add(BuildDependenciesOnlyFileCollectionResolveContext.java:83)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.visitContents(DefaultConfigurableFileCollection.java:91)
	at org.gradle.api.internal.file.CompositeFileCollection.visitDependencies(CompositeFileCollection.java:156)
	at org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection.visitDependencies(DefaultConfigurableFileCollection.java:97)
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext$TaskGraphImpl.getNodeValues(CachingTaskDependencyResolveContext.java:88)
	at org.gradle.internal.graph.CachingDirectedGraphWalker$GraphWithEmpyEdges.getNodeValues(CachingDirectedGraphWalker.java:202)
	at org.gradle.internal.graph.CachingDirectedGraphWalker.doSearch(CachingDirectedGraphWalker.java:112)
	at org.gradle.internal.graph.CachingDirectedGraphWalker.findValues(CachingDirectedGraphWalker.java:64)
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.doResolve(CachingTaskDependencyResolveContext.java:76)
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.resolve(CachingTaskDependencyResolveContext.java:65)
	... 41 more
```

**Comment by eljobe** [comment]
@Myllyenko It seems to me and @big-guy (we've talked about it off-thread) that `overview` really shouldn't have been a `String` in the first place. Here's our suggestion.

1. Change the overview field to be a [TextResource](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/groovy/org/gradle/api/resources/TextResource.java) and Annotate it with both `@Optional` and `@Nested`
1. Update the implementations of `String getOverview()` and `void setOverview(String overview)` to convert between the `String` as a path and a `TextResource`
1. Add two new methods `TextResource getOverviewText()` and `void setOverviewText(TextResource overviewText)` which should be the preferred way of getting and setting this field.
1. Deprecate the `getOverview()` and `setOverview()` methods.

This should get everything rolling again. Let us know if you need more guidance.

If you need an example of how to deprecate fields, you can take a look at the recent changes to the [ScalaCompileOptions.isAnt() method](https://github.com/gradle/gradle/blob/master/subprojects/scala/src/main/groovy/org/gradle/api/tasks/scala/ScalaCompileOptions.java#L60)

**Comment by Myllyenko** [comment]
Thank you for your help!
I've made these changes and slightly extended the new test method.

File: subprojects/language-groovy/src/main/java/org/gradle/api/tasks/javadoc/Groovydoc.java [hunk 1]
```diff
@@ -268,16 +269,42 @@ public void setFooter(String footer) {
 
     /**
      * Returns a HTML file to be used for overview documentation. Set to {@code null} when there is no overview file.
+     *
+     * @deprecated please use {@link #getOverviewText()} instead
      */
+    @Deprecated
```
Comment (eljobe) [inline subprojects/language-groovy/src/main/java/org/gradle/api/tasks/javadoc/Groovydoc.java]: In addition to marking the method as deprecated, you also need to use the SingleMessageLogger.nagUserOfDeprecated() method [like in the sample I mentioned](https://github.com/gradle/gradle/blob/master/subprojects/scala/src/main/groovy/org/gradle/api/tasks/scala/ScalaCompileOptions.java#L62).

This is also true for the setter.

**Comment by Myllyenko** [comment]
Done.

By the way, I think the similiar changes should be made for the Javadoc task: https://github.com/gradle/gradle/blob/master/subprojects/language-java/src/main/java/org/gradle/external/javadoc/MinimalJavadocOptions.java#L33

Am i right?

**Comment by eljobe** [comment]
@Myllyenko I agree about making the change for javadoc as well. But, let's do that in a separate change.

**Comment by eljobe** [comment]
@Myllyenko Thanks for the contribution to Groovydoc generation. Your work eliminates an annoying bug in the up-to-date checks that we'll all be happier not to have around any more.

We'd love to see something similar from you for the javadoc work,
Pepper

**Comment by eljobe** [comment]
We're trying out a new process for reviewing pull requests which calls for including our implementation check-list in the pull request itself. Please bear with us as we are experimenting.

The review for the Pull Request is at: https://code-review.gradle.org/cru/REVIEW-5897

- [ ] Does the change work?
  - [ ] Try out the feature, executing it like a end-user would.
  - [ ] Check corner cases and try to break it in various ways.
  - [ ] Are the error messages informative?
- [ ] Is the change sufficiently tested?
  - [ ] Are there appropriate levels of unit & integration tests?
  - [ ] Do the tests indicate what they are actually testing?
  - [ ] Are the tests in the right location?
  - [ ] Can new tests be consolidated with any existing tests?
- [ ] Does it change the public API?
  - [ ] Has @Incubating been used appropriately?
  - [ ] Does it preserve backward compatibility?
- [ ] Is it sufficiently documented?
  - [ ] Is the user guide documentation in the best location?
  - [ ] Is the user guide documentation clear and informative?
  - [ ] If there should be samples, are they helpful and tested appropriately?
  - [ ] Is there sufficient API/DSL documentation?
  - [ ] Is the API/DSL documentation clear and informative?
  - [ ] Do the release notes clearly convey the value of the feature?
  - [ ] Do the release notes explain the feature and link to more documentation?
- [ ] Is it implemented well?
  - [ ] Does it minimize surface area (particularly public)?
  - [ ] Does it follow our coding conventions? (note that we don't have coding conventions formalized but it's something to keep in mind)
  - [ ] Does it use appropriate data structures?
  - [ ] Does it minimize potential for invalid states?
  - [ ] Does it suitably reuse existing infrastructure?
  - [ ] Does it suitably consider concurrency?
  - [ ] Does it suitably consider failure modes?
  - [ ] Does it suitably consider performance?
  - [ ] Does it introduce new types or move existing ones? Are they located in the appropriate package? (e.g. internal classes in internal packages)

**Comment by Myllyenko** [comment]
> @Myllyenko Thanks for the contribution to Groovydoc generation.

Thank you too for the help!

> The review for the Pull Request is at: https://code-review.gradle.org/cru/REVIEW-5897

I have no access to this link. I've created an account on https://issues.gradle.org but it doesn't work on https://code-review.gradle.org.

**Comment by eljobe** [comment]
@Myllyenko I was worried that including that link would cause this confusion. That is for the internal review of this change. Only the gradle core team has access. It is basically just for cleaning up anything we might have overlooked during the initial merge. You, as the initial contributor, don't need to worry about that review. Sorry for the confusion.

**Comment by Myllyenko** [comment]
OK :)

**Pull request merged**